File: docs/neo-play-design-system/adapters/pyqt6/theme_tokens.py
"""
Neo Play design tokens — PyQt6 constants module.

Auto-derived from tokens/tokens.json. Import these instead of hardcoding
hex colors, px spacing, radii, or font specs anywhere in the app.

Usage:
    from theme_tokens import Color, Space, Radius, Type, Elevation
    widget.setStyleSheet(f"background-color: {Color.PRIMARY}; border-radius: {Radius.SM}px;")
"""

from dataclasses import dataclass
from PyQt6.QtGui import QColor, QFont
from PyQt6.QtWidgets import QGraphicsDropShadowEffect


class Color:
    """Hex color tokens. Wrap in QColor(...) where a QColor object is needed."""
    PRIMARY = "#02D962"
    ON_PRIMARY = "#FFFFFF"
    PRIMARY_CONTAINER = "#C6F7D8"
    ON_PRIMARY_CONTAINER = "#00210F"

    SECONDARY = "#2E77E7"
    ON_SECONDARY = "#FFFFFF"
    SECONDARY_CONTAINER = "#D6E7FF"
    ON_SECONDARY_CONTAINER = "#001C3B"

    TERTIARY = "#FE2151"
    ON_TERTIARY = "#FFFFFF"

    ERROR = "#B3261E"
    ON_ERROR = "#FFFFFF"

    SURFACE = "#FFFFFF"
    ON_SURFACE = "#1A1C1E"
    SURFACE_VARIANT = "#E1E2E8"
    OUTLINE = "#74777F"

    @staticmethod
    def q(hex_value: str) -> QColor:
        """Convenience: get a QColor from any hex token above."""
        return QColor(hex_value)


class Space:
    """Spacing scale in px (use directly as ints for margins/padding/spacing)."""
    XS = 4
    SM = 8
    MD = 16
    LG = 24
    XL = 32
    XXL = 48


class Radius:
    """Corner radius scale in px."""
    NONE = 0
    XS = 4
    SM = 8
    MD = 12
    LG = 16
    XL = 28
    FULL = 9999  # use min(width, height) / 2 in practice for pill/circular widgets


@dataclass
class TypeStyle:
    size: int
    line_height: int
    weight: int


class Type:
    """
    Typography scale. `size` maps to QFont.setPointSize/setPixelSize (use pixel size
    to match the design spec exactly); `weight` maps to QFont.Weight buckets below.
    """
    FONT_FAMILY = "Roboto"  # fallback handled by Qt font substitution if not installed

    DISPLAY_LARGE = TypeStyle(size=57, line_height=64, weight=400)
    DISPLAY_MEDIUM = TypeStyle(size=45, line_height=52, weight=400)
    HEADLINE_LARGE = TypeStyle(size=32, line_height=40, weight=400)
    HEADLINE_MEDIUM = TypeStyle(size=28, line_height=36, weight=400)
    TITLE_LARGE = TypeStyle(size=22, line_height=28, weight=500)
    TITLE_MEDIUM = TypeStyle(size=16, line_height=24, weight=500)
    TITLE_SMALL = TypeStyle(size=14, line_height=20, weight=500)
    BODY_LARGE = TypeStyle(size=16, line_height=24, weight=400)
    BODY_MEDIUM = TypeStyle(size=14, line_height=20, weight=400)
    BODY_SMALL = TypeStyle(size=12, line_height=16, weight=400)
    LABEL_LARGE = TypeStyle(size=14, line_height=20, weight=500)
    LABEL_MEDIUM = TypeStyle(size=12, line_height=16, weight=500)
    LABEL_SMALL = TypeStyle(size=11, line_height=16, weight=500)

    @staticmethod
    def qfont(style: TypeStyle) -> QFont:
        """Build a QFont from a TypeStyle. Weight 500+ maps to QFont.Weight.Medium/DemiBold."""
        font = QFont(Type.FONT_FAMILY)
        font.setPixelSize(style.size)
        if style.weight >= 700:
            font.setWeight(QFont.Weight.Bold)
        elif style.weight >= 500:
            font.setWeight(QFont.Weight.Medium)
        else:
            font.setWeight(QFont.Weight.Normal)
        return font


class Elevation:
    """
    Elevation levels. QSS has no real box-shadow support, so use
    QGraphicsDropShadowEffect on the widget instead (see `effect()` below).
    (blur_radius, x_offset, y_offset, alpha 0-255)
    """
    LEVELS = {
        0: (0, 0, 0, 0),
        1: (4, 0, 1, 20),
        2: (6, 0, 1, 30),
        3: (12, 0, 4, 36),
        4: (16, 0, 6, 40),
        5: (20, 0, 8, 46),
    }

    @staticmethod
    def effect(level: int) -> QGraphicsDropShadowEffect:
        """Return a configured QGraphicsDropShadowEffect for the given elevation level (0-5)."""
        blur, x, y, alpha = Elevation.LEVELS.get(level, Elevation.LEVELS[0])
        effect = QGraphicsDropShadowEffect()
        effect.setBlurRadius(blur)
        effect.setXOffset(x)
        effect.setYOffset(y)
        effect.setColor(QColor(0, 0, 0, alpha))
        return effect

File: docs/neo-play-design-system/adapters/gtk-rust/theme_tokens.rs
//! Neo Play design tokens — Rust constants module for gtk-rs.
//!
//! Auto-derived from tokens/tokens.json. Use these instead of hardcoding
//! hex strings, px values, or font specs anywhere in application code.
//! Colors are also defined in `theme.css` via `@define-color` for use in
//! GTK CSS selectors — keep both in sync if a token value changes.
//!
//! Usage:
//!     use theme_tokens::{Color, Space, Radius, Elevation, TypeStyle, Type};
//!     let rgba = Color::PRIMARY.to_rgba();

/// Hex color tokens, plus an RGBA helper for gtk-rs/gdk.
pub mod color {
    pub struct Hex(pub &'static str);

    impl Hex {
        /// Parse this token into a `gdk::RGBA`. Requires the `gdk` crate.
        pub fn to_rgba(&self) -> gdk::RGBA {
            gdk::RGBA::parse(self.0).expect("invalid token hex color")
        }
    }

    pub const PRIMARY: Hex = Hex("#02D962");
    pub const ON_PRIMARY: Hex = Hex("#FFFFFF");
    pub const PRIMARY_CONTAINER: Hex = Hex("#C6F7D8");
    pub const ON_PRIMARY_CONTAINER: Hex = Hex("#00210F");

    pub const SECONDARY: Hex = Hex("#2E77E7");
    pub const ON_SECONDARY: Hex = Hex("#FFFFFF");
    pub const SECONDARY_CONTAINER: Hex = Hex("#D6E7FF");
    pub const ON_SECONDARY_CONTAINER: Hex = Hex("#001C3B");

    pub const TERTIARY: Hex = Hex("#FE2151");
    pub const ON_TERTIARY: Hex = Hex("#FFFFFF");

    pub const ERROR: Hex = Hex("#B3261E");
    pub const ON_ERROR: Hex = Hex("#FFFFFF");

    pub const SURFACE: Hex = Hex("#FFFFFF");
    pub const ON_SURFACE: Hex = Hex("#1A1C1E");
    pub const SURFACE_VARIANT: Hex = Hex("#E1E2E8");
    pub const OUTLINE: Hex = Hex("#74777F");
}

/// Spacing scale, in px (i32 — use directly for GTK margins/spacing setters).
pub mod space {
    pub const XS: i32 = 4;
    pub const SM: i32 = 8;
    pub const MD: i32 = 16;
    pub const LG: i32 = 24;
    pub const XL: i32 = 32;
    pub const XXL: i32 = 48;
}

/// Corner radius scale, in px.
pub mod radius {
    pub const NONE: i32 = 0;
    pub const XS: i32 = 4;
    pub const SM: i32 = 8;
    pub const MD: i32 = 12;
    pub const LG: i32 = 16;
    pub const XL: i32 = 28;
    /// Not a literal usable radius — for true pill/circular widgets,
    /// use `min(width, height) / 2` at runtime instead.
    pub const FULL_SENTINEL: i32 = 9999;
}

/// Typography scale. Pair with a `pango::FontDescription` in gtk-rs.
pub mod typography {
    #[derive(Clone, Copy)]
    pub struct TypeStyle {
        pub size_px: i32,
        pub line_height_px: i32,
        pub weight: i32, // maps to pango::Weight buckets, e.g. 400 = Normal, 500 = Medium, 700 = Bold
    }

    pub const FONT_FAMILY: &str = "Roboto";

    pub const DISPLAY_LARGE: TypeStyle = TypeStyle { size_px: 57, line_height_px: 64, weight: 400 };
    pub const DISPLAY_MEDIUM: TypeStyle = TypeStyle { size_px: 45, line_height_px: 52, weight: 400 };
    pub const HEADLINE_LARGE: TypeStyle = TypeStyle { size_px: 32, line_height_px: 40, weight: 400 };
    pub const HEADLINE_MEDIUM: TypeStyle = TypeStyle { size_px: 28, line_height_px: 36, weight: 400 };
    pub const TITLE_LARGE: TypeStyle = TypeStyle { size_px: 22, line_height_px: 28, weight: 500 };
    pub const TITLE_MEDIUM: TypeStyle = TypeStyle { size_px: 16, line_height_px: 24, weight: 500 };
    pub const TITLE_SMALL: TypeStyle = TypeStyle { size_px: 14, line_height_px: 20, weight: 500 };
    pub const BODY_LARGE: TypeStyle = TypeStyle { size_px: 16, line_height_px: 24, weight: 400 };
    pub const BODY_MEDIUM: TypeStyle = TypeStyle { size_px: 14, line_height_px: 20, weight: 400 };
    pub const BODY_SMALL: TypeStyle = TypeStyle { size_px: 12, line_height_px: 16, weight: 400 };
    pub const LABEL_LARGE: TypeStyle = TypeStyle { size_px: 14, line_height_px: 20, weight: 500 };
    pub const LABEL_MEDIUM: TypeStyle = TypeStyle { size_px: 12, line_height_px: 16, weight: 500 };
    pub const LABEL_SMALL: TypeStyle = TypeStyle { size_px: 11, line_height_px: 16, weight: 500 };

    /// Build a `pango::FontDescription` from a TypeStyle (requires the `pango` crate).
    pub fn font_description(style: TypeStyle) -> pango::FontDescription {
        let mut desc = pango::FontDescription::new();
        desc.set_family(FONT_FAMILY);
        desc.set_size(style.size_px * pango::SCALE);
        desc.set_weight(match style.weight {
            w if w >= 700 => pango::Weight::Bold,
            w if w >= 500 => pango::Weight::Medium,
            _ => pango::Weight::Normal,
        });
        desc
    }
}

/// Elevation levels, expressed as (blur_px, x_offset, y_offset, alpha 0.0-1.0)
/// matching the box-shadow values already applied via theme.css. Provided here
/// mainly for cases where a shadow needs to be drawn manually (e.g. custom
/// GtkDrawingArea/Snapshot rendering) rather than through CSS.
pub mod elevation {
    pub struct Shadow {
        pub blur_px: f64,
        pub x_offset: f64,
        pub y_offset: f64,
        pub alpha: f64,
    }

    pub const LEVEL_0: Shadow = Shadow { blur_px: 0.0, x_offset: 0.0, y_offset: 0.0, alpha: 0.0 };
    pub const LEVEL_1: Shadow = Shadow { blur_px: 2.0, x_offset: 0.0, y_offset: 1.0, alpha: 0.08 };
    pub const LEVEL_2: Shadow = Shadow { blur_px: 3.0, x_offset: 0.0, y_offset: 1.0, alpha: 0.12 };
    pub const LEVEL_3: Shadow = Shadow { blur_px: 8.0, x_offset: 0.0, y_offset: 4.0, alpha: 0.14 };
    pub const LEVEL_4: Shadow = Shadow { blur_px: 10.0, x_offset: 0.0, y_offset: 6.0, alpha: 0.16 };
    pub const LEVEL_5: Shadow = Shadow { blur_px: 12.0, x_offset: 0.0, y_offset: 8.0, alpha: 0.18 };
}
